### PR TITLE
feature(pu): add NGU algorithm 

### PR DESCRIPTION
**Pull request by puyuan1996**
Add exploration-RL algorithm NGU. Note that NGU algorithm has not been completed yet, it will being updated continuously.

### CODE REVIEW

File: ding/reward_model/rnd_reward_model.py [hunk 1]
```diff
@@ -69,8 +92,14 @@ def __init__(self, config: EasyDict, device: str, tb_logger: 'SummaryWriter') ->
         assert self.intrinsic_reward_type in ['add', 'new', 'assign']
         self.train_data = []
         self.opt = optim.Adam(self.reward_model.predictor.parameters(), config.learning_rate)
+        self.estimate_cnt = 0
 
     def _train(self) -> None:
+        if isinstance(self.train_data[0], list):  # if self.train_data list( list(torch.tensor) ) rnn
+            tmp = []
```
Comment (PaParaZz1) [inline ding/reward_model/rnd_reward_model.py:99]: just use `train_data = sum(train_data, [])`

File: ding/reward_model/rnd_reward_model.py [hunk 2]
```diff
@@ -89,20 +118,55 @@ def estimate(self, data: list) -> None:
...
+                    elif self.intrinsic_reward_type == 'new':
+                        item['intrinsic_reward'] = rew
+                    elif self.intrinsic_reward_type == 'assign':
+                        item['reward'] = rew
+            else:  #rnn nstep
+                reward = reward.to(data[0]['reward'][0].device)
+                reward = torch.chunk(reward, int(reward.shape[0]), dim=0)
+                # reward = torch.chunk(reward, int(reward.shape[0]/len(data[0]['reward'])), dim=0)
+                # reward.shape[0] 64 batch_size len(data[0]['reward'])) 24=20+2*2 eps_len
+                batch_size = int(reward.__len__() / len(data[0]['reward']))
```
Comment (PaParaZz1) [inline ding/reward_model/rnd_reward_model.py:149]: len(reward)

File: ding/reward_model/rnd_reward_model.py [hunk 3]
```diff
@@ -89,20 +118,55 @@ def estimate(self, data: list) -> None:
...
+                eps_len = len(data[0]['reward'])  ## pu if r2d2 n-step reward   fake_reward = torch.zeros(1)
+                for i in range(batch_size):  #64 batch_size
+                    for j in range(eps_len):  #24 24=20+2*2 eps_len
+                        if j < eps_len - self.cfg.nstep:
+                            bonus = torch.cat([reward[i * eps_len + j + k] for k in range(self.cfg.nstep)], dim=0)
+                            if self.intrinsic_reward_type == 'add':
+                                data[i]['reward'][j] += bonus
+                            # elif self.intrinsic_reward_type == 'new':
```
Comment (PaParaZz1) [inline ding/reward_model/rnd_reward_model.py:157]: Why comments `new` branch here?

File: ding/model/wrapper/model_wrappers.py [hunk 4]
```diff
@@ -240,6 +240,44 @@ def forward(self, *args, **kwargs):
...
+    Overview:
+        eps is a dict n_env
+        Epsilon greedy sampler used in collector_model to help balance exploratin and exploitation.
+    Interfaces:
+        register
+    """
+
+    def forward(self, *args, **kwargs):
+        kwargs.pop('eps')
+        eps ={i: 0.4 ** ( 1 + 8*i/(args[0]['obs'].shape[0]-1)) for i in range(args[0]['obs'].shape[0])} # TODO
```
Comment (PaParaZz1) [inline ding/model/wrapper/model_wrappers.py:254]: I think `eps` can be figured out outside and pass into here, so there is no need to add a new wrapper

File: ding/reward_model/rnd_reward_model.py [hunk 5]
```diff
@@ -19,7 +19,30 @@ def collect_states(iterator):  # get total_states
         res.append(state)
     return res
 
+class RndNetwork(nn.Module):
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply ding/reward_model/rnd_reward_model.py:22]: 这个rnd_reward_model.py文件是之前遗弃的版本，不好意思上次commit时没有untrack，我在下次commit时删除它

File: ding/torch_utils/data_helper.py [hunk 6]
```diff
@@ -8,7 +8,32 @@
 import numpy as np
 import torch
 
+def one_hot_embedding(labels, num_classes):
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply ding/torch_utils/data_helper.py:11]: 这里如果是一个episode开始时，action的one-hot会设置为全零，在data里这个动作是存为int(action_shape)的，可能需要下面的one_hot_embedding_none()来特殊处理

File: ding/reward_model/rnd_reward_model.py (hunk 1 above)
Comment (puyuan1996) [reply ding/reward_model/rnd_reward_model.py:99]: 这个也和上面一样，rnd_reward_model.py是遗弃了的。不过学到了这个list处理方式，谢谢啦

File: ding/model/wrapper/model_wrappers.py (hunk 4 above)
Comment (puyuan1996) [reply ding/model/wrapper/model_wrappers.py:255]: 好的我试着这样改一下

File: ding/entry/serial_entry_reward_model.py [hunk 7]
```diff
@@ -102,11 +104,11 @@ def serial_pipeline_reward_model(
             new_data = collector.collect(train_iter=learner.train_iter, policy_kwargs=collect_kwargs)
             new_data_count += len(new_data)
             # collect data for reward_model training
-            reward_model.collect_data(new_data)
+            # reward_model.collect_data(new_data) #TODO
             replay_buffer.push(new_data, cur_collector_envstep=collector.envstep)
         # update reward_model
-        reward_model.train()
-        reward_model.clear_data()
+        # reward_model.train()  # pu TODO
```
[the remark this answers is not in the document]
Comment (puyuan1996) [reply ding/entry/serial_entry_reward_model.py:110]: 好的

File: ding/torch_utils/data_helper.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PaParaZz1) [reply ding/torch_utils/data_helper.py:11]: OK, for this case, I think you can add this part into the original `one_hot` function. And I recommend use `-1` rather than `None` to label this special element, it can be arranged into a tensor.

[the remark this answers is not in the document]
Comment (puyuan1996) [reply ding/torch_utils/data_helper.py:11]: 好的

**Review by PaParaZz1 — APPROVED** [review]

**Comment by GoingMyWay** [comment]
> > Hi, is NGU now ready to use?
> 
> Hi,
> 
> Thank you very much for your question! Yes, our implemented [NGU](https://github.com/opendilab/DI-engine/blob/main/ding/policy/ngu.py) algorithm is ready to use.
> 
> * We tested it on the [minigrid](https://github.com/Farama-Foundation/MiniGrid) environment. In the hard-exploration env MiniGrid-DoorKey-16x16-v0, our implemented NGU algorithm converged at about 30M  env steps, which is aligned with the results in the unpublished [paper](https://openreview.net/forum?id=_ptUyYP19mP).
>   The learning curve for 2 seeds is as follows:
>   image
> * You can run this entry config https://github.com/opendilab/DI-engine/blob/main/dizoo/minigrid/config/minigrid_ngu_config.py to train NGU agent in MiniGrid env.
> * For the algorithm details of NGU, you can refer to our Chinese [blog](https://zhuanlan.zhihu.com/p/551992517). For more paper information about exploration in RL, you can refer to our [awesome-exploration-rl repo](https://github.com/opendilab/awesome-exploration-rl).
> 
> Good Luck!

@puyuan1996, Great thank you 蒲源.

**Comment by GoingMyWay** [comment]
> > Hi, is NGU now ready to use?
> 
> Hi,
> 
> Thank you very much for your question! Yes, our implemented [NGU](https://github.com/opendilab/DI-engine/blob/main/ding/policy/ngu.py) algorithm is ready to use.
> 
> * We tested it on the [minigrid](https://github.com/Farama-Foundation/MiniGrid) environment. In the hard-exploration env MiniGrid-DoorKey-16x16-v0, our implemented NGU algorithm converged at about 30M  env steps, which is aligned with the results in the unpublished [paper](https://openreview.net/forum?id=_ptUyYP19mP).
>   The learning curve for 2 seeds is as follows:
>   image
> * You can run this entry config https://github.com/opendilab/DI-engine/blob/main/dizoo/minigrid/config/minigrid_ngu_config.py to train NGU agent in MiniGrid env.
> * For the algorithm details of NGU, you can refer to our Chinese [blog](https://zhuanlan.zhihu.com/p/551992517). For more paper information about exploration in RL, you can refer to our [awesome-exploration-rl repo](https://github.com/opendilab/awesome-exploration-rl).
> 
> Good Luck!

@puyuan1996, Hi, did you run experiments on Atari hard exploration environments, such as Montezuma Revenge?
